tsfc/driver.py

```diff
@@ -363,6 +363,118 @@ def compile_expression_at_points(expression, points, coordinates, interface=None
+
+    :arg expression: UFL expression
+    :arg dual_functionals: FIAT dual functionals
+    :arg coordinates: the coordinate function
+    :arg interface: backend module for the kernel interface
+    :arg parameters: parameters object
+    :arg coffee: compile coffee kernel instead of loopy kernel
+    """
+    import coffee.base as ast
+    import loopy as lp
+
+    if parameters is None:
+        parameters = default_parameters()
+    else:
+        _ = default_parameters()
+        _.update(parameters)
+        parameters = _
+
+    # Determine whether in complex mode
+    complex_mode = is_complex(parameters["scalar_type"])
+
+    # Apply UFL preprocessing
+    expression = ufl_utils.preprocess_expression(expression,
+                                                 complex_mode=complex_mode)
+
+    # Initialise kernel builder
+    if interface is None:
+        if coffee:
+            import tsfc.kernel_interface.firedrake as firedrake_interface_coffee
+            interface = firedrake_interface_coffee.ExpressionKernelBuilder
+        else:
+            # Delayed import, loopy is a runtime dependency
+            import tsfc.kernel_interface.firedrake_loopy as firedrake_interface_loopy
+            interface = firedrake_interface_loopy.ExpressionKernelBuilder
+
+    builder = interface(parameters["scalar_type"])
+    arguments = extract_arguments(expression)
+    argument_multiindices = tuple(builder.create_element(arg.ufl_element()).get_indices()
+                                  for arg in arguments)
+
+    # Replace coordinates (if any)
+    domain = expression.ufl_domain()
+    if domain:
+        assert coordinates.ufl_domain() == domain
+        builder.domain_coordinate[domain] = coordinates
+        builder.set_cell_sizes(domain)
+
+    # Collect required coefficients
+    coefficients = extract_coefficients(expression)
+    if has_type(expression, GeometricQuantity) or any(fem.needs_coordinate_mapping(c.ufl_element()) for c in coefficients):
+        coefficients = [coordinates] + coefficients
+    builder.set_coefficients(coefficients)
+
+    # Split mixed coefficients
+    expression = ufl_utils.split_coefficients(expression, builder.coefficient_split)
+
+    # Translate to GEM
+    kernel_cfg = dict(interface=builder,
+                     ufl_cell=coordinates.ufl_domain().ufl_cell(),
+                     precision=parameters["precision"],
+                     integration_dim=coordinates.ufl_domain().ufl_cell().topological_dimension(),
+                     argument_multiindices=argument_multiindices)
+
+    dual_expressions = []
+    for dual in dual_functionals:
+        # Extract associated quadrature rule representation
+        # Make a quad_rule
+
+        quad_points = [*dual.pt_dict.keys()]
+        # get it working for scalar functions first; how do we deal with components??
+        weights = [dual.pt_dict[pt][0][0] for pt in quad_points]
+        quad_rule = QuadratureRule(PointSet(quad_points), weights)
+
```

Comment:
I would assert that dual.deriv_dict is empty here until it is figured out how to address that.  (One can always push the burden back on FIAT to expand derivative-based functionals into linear combinations of point values and then have only a minor change here)

Reply:
Fixed in later comment I hadn't gotten too.